path_following_controller/src/path_following_controller.cpp

```diff
@@ -0,0 +1,1145 @@
+//
+// Unauthorized copying of this file, via any medium is strictly prohibited.
+// The file is considered confidential.
+
+#include "path_following_controller/path_following_controller.hpp"
+
+#include <limits>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "angles/angles.h"
+#include "rclcpp_action/create_server.hpp"
+#include "rclcpp_action/server_goal_handle.hpp"
+#include "controller_interface/helpers.hpp"
+
+#include "lifecycle_msgs/msg/state.hpp"
+
+namespace path_following_controller
+{
+PathFollowingController::PathFollowingController() 
+: controller_interface::ControllerInterface() {}
+
+controller_interface::CallbackReturn PathFollowingController::on_init()
+{
+  try
+  {
+    // Create the parameter listener and get the parameters
+    param_listener_ = std::make_shared<ParamListener>(get_node());
+    params_ = param_listener_->get_params();
+
+  /*   TODO
+    // Set interpolation method from string parameter
+    interpolation_method_ = interpolation_methods::from_string(params_.interpolation_method);
+  */
+  }
+  catch (const std::exception & e)
+  {
+    fprintf(stderr, "Exception thrown during init stage with message: %s \n", e.what());
+    return CallbackReturn::ERROR;
+  }
+
+  /*   REVIEW NECESSITY
+  // TODO(christophfroehlich): remove deprecation warning
+  if (params_.allow_nonzero_velocity_at_trajectory_end)
+  {
+    RCLCPP_WARN(
+      get_node()->get_logger(),
+      "[Deprecated]: \"allow_nonzero_velocity_at_trajectory_end\" is set to "
+      "true. The default behavior will change to false.");
+  }
+  */
+
+  return CallbackReturn::SUCCESS;
+}
+
+controller_interface::return_type PathFollowingController::update(
+  const rclcpp::Time & time, const rclcpp::Duration & /*period*/)
+{
+  if (get_state().id() == lifecycle_msgs::msg::State::PRIMARY_STATE_INACTIVE)
+    return controller_interface::return_type::OK;
+  
+  // update dynamic parameters
+  if (param_listener_->is_old(params_))
+  {
+    params_ = param_listener_->get_params();
+    default_tolerances_ = get_segment_tolerances(params_);
+  }
+
+  ///////////////// GET OR REPLACE CURRENT GOAL
+  // get active goal from the action server
+  const auto active_goal = *rt_active_goal_.readFromRT();
+
+  // REVIEW: rewrite this logic in a nicer way, get rid of dereferencing
+  // check if there is a new goal message
+  auto current_external_msg = traj_external_point_ptr_->get_trajectory_msg();
+  auto new_external_msg = traj_msg_external_point_ptr_.readFromRT();    
+
+  // Discard action goal if it is pending, but not active (still somewhere in goal_handle_timer_)
+  // or if there was a new goal received, but from topic
+  if(
+    current_external_msg != *new_external_msg 
+    && (*(rt_action_goal_pending_.readFromRT()) && !active_goal) == false)
+  {
+    fill_partial_goal(*new_external_msg);
+    sort_to_local_joint_order(*new_external_msg);
+    traj_external_point_ptr_->update(*new_external_msg);   // resets trajectory index to 0
```

Comment:
I don't think that comment makes sense, or it is at last confusing. We are replacing here trajectory and this is the main goal.
```suggestion
    traj_external_point_ptr_->update(*new_external_msg);
```

Reply:
will remove